subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -176,6 +255,179 @@ static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+			       settings_read_cb read_cb, void *cb_arg)
+{
+	ssize_t len;
+	int rc;
+
+	if (ctx.buffer != NULL && ctx.cur_image_no == IMAGE_NO_FIXED_HEADER
+	    && strcmp(key, CBOR_HEADER_SETTING_NAME) == 0) {
+		if (len_rd > (ctx.buffer_size - FIXED_HEADER_SIZE)) {
+			LOG_ERR("CBOR header in settings is too large");
+			return -ENOMEM;
+		}
+
+		/* Restore the cbor header from the NVS settings if available */
+		len = read_cb(cb_arg, ctx.buffer + FIXED_HEADER_SIZE,
+			      ctx.buffer_size - FIXED_HEADER_SIZE);
+		if (len < 0) {
+			LOG_ERR("Can't read cbor header from storage");
+			return len;
+		}
+
+		ctx.cur_item_size = FIXED_HEADER_SIZE + len;
+		ctx.cur_image_no = IMAGE_NO_CBOR_HEADER;
+
+		rc = parse_cbor_header();
+
+		if (rc < 0) {
+			LOG_ERR("Failed to parse cbor header loaded from storage: %d", rc);
+			return rc;
+		}
+
+		ctx.cur_offset = ctx.cur_item_size;
+
+		select_next_image();
+	}
+
+	if (ctx.buffer != NULL && strcmp(key, IMAGE_FINISHED_SETTING_NAME) == 0) {
+		/* Restore the cbor header from the NVS settings if available */
+		uint8_t finished_image_no = 0;
+
+		if (len_rd != sizeof(finished_image_no)) {
+			/* Should never get here */
+			return -EINVAL;
+		}
+
+		len = read_cb(cb_arg, &finished_image_no, sizeof(finished_image_no));
+
+		if (len < 0) {
+			LOG_ERR("Can't read finished image number from storage");
+			return len;
+		}
+
+		if (ctx.max_loaded_finished_image_no < finished_image_no) {
+			ctx.max_loaded_finished_image_no = finished_image_no;
+		}
+	}
+
+	return 0;
+}
+
+static int load_saved_progress(void)
+{
+	int err = 0;
+	static struct settings_handler sh = {
+		.name = SETTINGS_MODULE,
+		.h_set = module_settings_set,
+	};
+
+	LOG_INF("Loading saved progress from NVM");
+
+	ctx.max_loaded_finished_image_no = -1;
+
+	/* settings_subsys_init is idempotent so this is safe to do. */
+	err = settings_subsys_init();
+	if (err) {
+		LOG_ERR("settings_subsys_init failed (err %d)", err);
+		return err;
+	}
+
+	err = settings_register(&sh);
+	if (err && err != -EEXIST) {
+		LOG_ERR("setting_register failed: (err %d)", err);
+		return err;
+	}
+
+	/* Load settings from NVM. If the update was interrupted, the context will be restored */
+	err = settings_load();
+	if (err) {
+		LOG_ERR("settings_load failed (err %d)", err);
+		return err;
+	}
+
+	if (ctx.cur_image_no == IMAGE_NO_FIXED_HEADER) {
+		LOG_INF("No saved progress found");
+		/* Progress was not stored in settings, start from the beginning */
+		ctx.saved_progress_loaded = true;
+		return 0;
+	}
+
+	if (ctx.cur_image_no < 0) {
+		/* Should never get here */
+		return -EFAULT;
+	}
+
+	while (ctx.cur_image_no < ctx.header.image_count) {
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (writer == NULL) {
+			LOG_WRN("No writer registered for image %d!", ctx.cur_image_no);
+			err = -EPIPE;
+			break;
+		}
+		if (writer->offset == NULL) {
+			LOG_WRN("No function to read offset registered for image %d!",
```

Comment:
Is there a need to get offset for images prior to `max_loaded_finished_image_no + 1`?